File: src/mojollama/mojo_engine/transpose_q8.c
/* transpose_q8.c — Transpose Q8_0 matrix files from [d0,d1] to [d1,d0].
   Reads a shape map: each line is "<filename>,<d0>,<d1>" (1D/2D).
   Usage: ./transpose_q8 <in_dir> <out_dir> <shapes.csv>
*/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <dirent.h>
#include <sys/stat.h>
#include <math.h>

#define QK 32
#define QB 34
#define MAX_PATH 4096
#define MAX_LINE 512

static uint16_t f32_to_f16(float f) {
    uint32_t u; memcpy(&u, &f, 4);
    int s = (u >> 31) & 1, e = (u >> 23) & 0xFF, m = u & 0x7FFFFF;
    if (e == 0) return s << 15;
    if (e == 0xFF) return (s << 15) | 0x7C00;
    int ne = e - 127 + 15;
    if (ne >= 31) return (s << 15) | 0x7C00;
    if (ne <= 0) return s << 15;
    return (s << 15) | (ne << 10) | (m >> 13);
}

/* Dequantize Q8_0 block (34 bytes → 32 floats) */
static void deq_q8_block(const uint8_t *blk, float *out) {
    uint16_t s_u; memcpy(&s_u, blk, 2);
    float scale;
    { int s=(s_u>>15)&1,e=(s_u>>10)&0x1F,m=s_u&0x3FF;
      if(e==0) scale=(float)m*5.96e-8f; else if(e==31) scale=0;
      else { uint32_t b=(s<<31)|((e+112)<<23)|(m<<13); memcpy(&scale,&b,4); } }
    for (int i = 0; i < QK; i++)
        out[i] = (float)((int8_t)blk[2+i]) * scale;
}

/* Quantize float32 row → Q8_0 bytes (appended to buf) */
static void q8_quant_row(const float *row, int n, uint8_t *buf, long *off) {
    int n_blk = (n + QK - 1) / QK;
    for (int b = 0; b < n_blk; b++) {
        int s = b * QK, e = s + QK < n ? s + QK : n;
        float max_abs = 0;
        for (int i = s; i < e; i++) { float a = fabsf(row[i]); if (a > max_abs) max_abs = a; }
        float scale = max_abs > 1e-10f ? max_abs / 127.0f : 1.0f;
        uint16_t su = f32_to_f16(scale);
        memcpy(buf + *off, &su, 2); *off += 2;
        for (int i = s; i < e; i++) {
            int qv = (int)roundf(row[i] / scale);
            if (qv > 127) qv = 127; if (qv < -128) qv = -128;
            buf[(*off)++] = (uint8_t)(qv & 0xFF);
        }
        // Pad to QK
        for (int i = e; i < s + QK; i++) buf[(*off)++] = 0;
    }
}

int main(int argc, char **argv) {
    if (argc < 4) {
        fprintf(stderr, "Usage: %s <in_dir> <out_dir> <shapes.csv>\n", argv[0]);
        fprintf(stderr, "  shapes.csv: each line 'filename,d0,d1' where d1=0 means 1D (copy)\n");
        return 1;
    }
    
    mkdir(argv[2], 0755);
    
    // Read shape map
    FILE *sf = fopen(argv[3], "r");
    if (!sf) { perror("shapes.csv"); return 1; }
    
    char line[MAX_LINE];
    int n_total = 0, n_transpose = 0;
    
    while (fgets(line, sizeof(line), sf)) {
        // Strip newline
        char *nl = strchr(line, '\n'); if (nl) *nl = 0;
        if (line[0] == 0) continue;
        
        char fname[MAX_PATH];
        long d0, d1;
        if (sscanf(line, "%[^,],%ld,%ld", fname, &d0, &d1) != 3) continue;
        
        char in_path[MAX_PATH], out_path[MAX_PATH];
        snprintf(in_path, sizeof(in_path), "%s/%s", argv[1], fname);
        snprintf(out_path, sizeof(out_path), "%s/%s", argv[2], fname);
        
        // Check if output already exists
        struct stat st;
        if (stat(out_path, &st) == 0) continue;  // Already processed
        
        // Check input exists
        if (stat(in_path, &st) != 0) {
            fprintf(stderr, "  SKIP: %s not found\n", fname);
            continue;
        }
        
        // Read input file
        FILE *f = fopen(in_path, "rb");
        if (!f) { fprintf(stderr, "  FAIL: cannot open %s\n", fname); continue; }
        fseek(f, 0, SEEK_END); long sz = ftell(f); fseek(f, 0, SEEK_SET);
        uint8_t *in_data = (uint8_t*)malloc(sz);
        if (!in_data) { fclose(f); fprintf(stderr, "  FAIL: malloc %s\n", fname); continue; }
        fread(in_data, 1, sz, f); fclose(f);
        
        if (d1 == 0 || d0 == d1) {
            // 1D or square: just copy
            FILE *out = fopen(out_path, "wb");
            if (out) { fwrite(in_data, 1, sz, out); fclose(out); }
            free(in_data);
            n_total++;
            if (n_total % 100 == 0) fprintf(stderr, "  copied %d files...\r", n_total);
            continue;
        }
        
        // Transpose: [d0, d1] → [d1, d0]
        // Dequantize entire matrix
        long n_vals = d0 * d1;
        float *f32 = (float*)malloc(n_vals * sizeof(float));
        if (!f32) { free(in_data); fprintf(stderr, "  FAIL: malloc f32 %s\n", fname); continue; }
        
        long n_q8_blk = sz / QB;
        long n_q8_elems = n_q8_blk * QK;
        // Ensure we don't read past data
        long n_read = n_q8_elems < n_vals ? n_q8_elems : n_vals;
        for (long b = 0; b < n_read / QK; b++)
            deq_q8_block(in_data + b * QB, f32 + b * QK);
        
        free(in_data);
        
        // Transpose: f32[d0][d1] → f32_T[d1][d0]
        float *f32_T = (float*)malloc(n_vals * sizeof(float));
        if (!f32_T) { free(f32); fprintf(stderr, "  FAIL: malloc T %s\n", fname); continue; }
        
        for (long r = 0; r < d0; r++)
            for (long c = 0; c < d1; c++)
                f32_T[c * d0 + r] = f32[r * d1 + c];
        
        free(f32);
        
        // Quantize back to Q8_0
        long out_sz = ((d1 * d0 + QK - 1) / QK) * QB;
        uint8_t *q8 = (uint8_t*)calloc(1, out_sz);
        if (!q8) { free(f32_T); fprintf(stderr, "  FAIL: malloc q8 %s\n", fname); continue; }
        
        long off = 0;
        for (long r = 0; r < d1; r++)
            q8_quant_row(f32_T + r * d0, d0, q8, &off);
        
        free(f32_T);
        
        FILE *out = fopen(out_path, "wb");
        if (out) { fwrite(q8, 1, out_sz, out); fclose(out); }
        free(q8);
        
        n_total++; n_transpose++;
        if (n_total % 50 == 0)
            fprintf(stderr, "  %d files (%d transposed)...\r", n_total, n_transpose);
    }
    fclose(sf);
    
    fprintf(stderr, "\nDone: %d total, %d transposed\n", n_total, n_transpose);
    return 0;
}

File: src/mojollama/mojo_engine/transpose_q8.py
#!/usr/bin/env python3
"""
Transpose Q8_0 weight files from GGUF orientation [d0, d1] to engine [d1, d0].
Usage: python3 transpose_q8.py <in_dir> <out_dir> <tensor_shapes.json>
  Where tensor_shapes.json maps filename -> [d0, d1]
"""
import os, sys, struct, json, numpy as np, time
from concurrent.futures import ProcessPoolExecutor, as_completed

QK, QB = 32, 34

def q8_read(fpath):
    """Read Q8_0 file, return (data, n_elements)."""
    with open(fpath, 'rb') as f:
        raw = f.read()
    n_blocks = len(raw) // QB
    n_elems = n_blocks * QK
    # Trim any trailing partial block (shouldn't happen for full files)
    return raw, n_elems

def q8_dequant_block(blk_bytes):
    """Dequantize a single Q8_0 block (QB=34 bytes) -> 32 float32 values."""
    scale = struct.unpack('<e', blk_bytes[:2])[0]
    quants = np.frombuffer(blk_bytes[2:], dtype=np.int8, count=QK)
    return quants.astype(np.float32) * scale

def dequant_q8(raw, n):
    """Dequantize entire Q8_0 file to float32."""
    n_blocks = len(raw) // QB
    # Process in chunks for efficiency
    f32 = np.zeros(n_blocks * QK, dtype=np.float32)
    for b in range(n_blocks):
        off = b * QB
        scale = struct.unpack('<e', raw[off:off+2])[0]
        quants = np.frombuffer(raw[off+2:off+QB], dtype=np.int8)
        f32[b*QK:(b+1)*QK] = quants.astype(np.float32) * scale
    return f32[:n]

def quant_q8_row(row):
    """Quantize a 1D float32 array to Q8_0 bytes."""
    n = len(row)
    n_blk = (n + QK - 1) // QK
    buf = bytearray()
    for b in range(n_blk):
        s = b * QK
        e = min(s + QK, n)
        block = row[s:e]
        max_abs = float(np.max(np.abs(block))) if len(block) > 0 else 0.0
        scale = max_abs / 127.0 if max_abs > 1e-10 else 1.0
        qvals = np.clip(np.round(block / scale), -128, 127).astype(np.int8)
        buf += struct.pack('<e', np.float16(scale))
        buf += qvals.tobytes()
        buf += b'\x00' * (QK - len(qvals))
    return bytes(buf)

def transpose_file(in_path, out_path, d0, d1):
    """Transpose a Q8_0 weight matrix from [d0, d1] to [d1, d0]."""
    raw, n = q8_read(in_path)
    n_actual = d0 * d1
    # Dequantize
    f32 = dequant_q8(raw, n)
    if len(f32) != n_actual:
        print(f"    WARNING: {os.path.basename(in_path)}: expected {n_actual} elements, got {len(f32)}")
        # Trim or pad
        f32 = f32[:n_actual] if len(f32) > n_actual else np.pad(f32, (0, n_actual - len(f32)))
    # Reshape and transpose
    mat = f32.reshape(d0, d1).T  # Now [d1, d0]
    # Requantize to Q8_0
    buf = bytearray()
    for r in range(d1):
        buf += quant_q8_row(mat[r])
    with open(out_path, 'wb') as f:
        f.write(buf)
    return True

def main():
    in_dir = sys.argv[1]
    out_dir = sys.argv[2]
    shape_file = sys.argv[3] if len(sys.argv) > 3 else None
    
    os.makedirs(out_dir, exist_ok=True)
    
    # Load shape mapping
    with open(shape_file) as f:
        shapes = json.load(f)  # dict: filename -> [d0, d1, need_transpose]
    
    t0 = time.time()
    n_total = len([v for v in shapes.values() if len(v) >= 3 and v[2]])
    n_done = 0
    
    for fname, info in sorted(shapes.items()):
        if len(info) < 3 or not info[2]:
            # No transpose needed, just copy
            in_path = os.path.join(in_dir, fname)
            out_path = os.path.join(out_dir, fname)
            if os.path.exists(in_path) and not os.path.exists(out_path):
                with open(in_path, 'rb') as f:
                    data = f.read()
                with open(out_path, 'wb') as f:
                    f.write(data)
            continue
        
        d0, d1 = info[0], info[1]
        in_path = os.path.join(in_dir, fname)
        out_path = os.path.join(out_dir, fname)
        
        if os.path.exists(out_path):
            continue  # Already done
        
        transpose_file(in_path, out_path, d0, d1)
        n_done += 1
        elapsed = time.time() - t0
        if n_done % 20 == 0:
            print(f"  [{n_done}/{n_total}] {fname} ({elapsed:.0f}s)", flush=True)
    
    # Copy remaining non-transposed files
    for fname in os.listdir(in_dir):
        if fname not in shapes:
            in_path = os.path.join(in_dir, fname)
            out_path = os.path.join(out_dir, fname)
            if os.path.exists(in_path) and not os.path.exists(out_path):
                with open(in_path, 'rb') as f:
                    with open(out_path, 'wb') as f2:
                        f2.write(f.read())
    
    print(f"\nDone: {n_done} transposed in {time.time()-t0:.0f}s", flush=True)

if __name__ == '__main__':
    main()
